renderdoc/serialise/codecs/vk_cpp_codec/intervals.h

```diff
@@ -0,0 +1,175 @@
+struct Intervals;
+
+template <typename T>
+struct IntervalsIter
+{
+  friend struct Intervals<T>;
+
+private:
+  typename std::map<uint64_t, T>::iterator iter;
+  std::map<uint64_t, T> *owner;
+  IntervalsIter(std::map<uint64_t, T> *owner, typename std::map<uint64_t, T>::iterator iter)
+      : iter(iter), owner(owner)
+  {
+  }
+  typename std::map<uint64_t, T>::iterator unwrap() { return iter; }
+public:
+  IntervalsIter(const IntervalsIter &src) : owner(src.owner), iter(src.iter) {}
+  IntervalsIter &operator++()
+  {
+    ++iter;
+    return *this;
+  }
+  IntervalsIter operator++(int)
+  {
+    IntervalsIter tmp(*this);
+    operator++();
+    return tmp;
+  }
+  IntervalsIter &operator--()
+  {
+    --iter;
+    return *this;
+  }
+  IntervalsIter operator--(int)
+  {
+    IntervalsIter tmp(*this);
+    operator--();
+    return tmp;
+  }
+  bool operator==(const IntervalsIter &rhs) const { return iter == rhs.iter && owner == rhs.owner; }
+  bool operator!=(const IntervalsIter &rhs) const { return iter != rhs.iter && owner == rhs.owner; }
+  IntervalsIter &operator=(const IntervalsIter &rhs)
+  {
+    iter = rhs.iter;
+    owner = rhs.owner;
+    return *this;
+  }
+  inline uint64_t start() const { return iter->first; }
+  inline uint64_t end() const
+  {
+    typename std::map<uint64_t, T>::iterator next = iter;
+    next++;
+    if(next == owner->end())
+    {
+      return UINT64_MAX;
+    }
+    return next->first;
+  }
+  inline const T &value() const { return iter->second; }
+  inline void setValue(uint64_t aStart, uint64_t aEnd, const T &aValue)
+  {
+    T old_value = this->value();
+    if((aValue == old_value || aEnd <= start()) && end() <= aStart)
+    {
+      // The value is unchanged, or the specified interval is disjoint from this interval.
+      return;
+    }
+
+    // Add a new endpoint for aStart, if necessary, and update iter's value to aValue.
+    if(start() < aStart)
+    {
+      // The updated portion of this interval begins at aStart.
+      iter = owner->insert(std::pair<uint64_t, T>(aStart, aValue)).first;
+    }
+    else
+    {
+      // The updated portion of this interval begins at start()
+      iter->second = aValue;
+    }
+
+    // Add a new endpoint for aEnd, if necessary.
+    if(aEnd < end())
+    {
+      owner->insert(std::pair<uint64_t, T>(aEnd, old_value));
+    }
+
+    // Merge with preceding interval, if necessary
+    if(iter != owner->begin())
+    {
+      typename std::map<uint64_t, T>::iterator prev_it = iter;
+      prev_it--;
+      if(prev_it->second == iter->second)
+      {
+        owner->erase(iter);
+        iter = prev_it;
+      }
+    }
+
+    // Merge with succeding interval, if necessary
+    typename std::map<uint64_t, T>::iterator next_it = iter;
+    next_it++;
+    if(next_it != owner->end())
+    {
+      if(next_it->second == iter->second)
+      {
+        owner->erase(next_it);
+      }
+    }
+  }
+};
+
+template <typename T>
+struct Intervals
```

Comment:
This data structure could be useful in other cases, so I think this might be better placed in the common code and not being in vk_cpp_codec specifically. E.g. it would be easy enough to add a `keytype` template parameter in future and use `ResourceId`.

Would you mind moving it to `renderdoc/core/intervals.h` and out of the namespace?

Reply:
Done.